source/lib.ts

```diff
@@ -0,0 +1,323 @@
+// /source/lib.ts
+// The option parser and rate limiting middleware
+
+import Express from 'express'
+
+import {
+	Options,
+	AugmentedRequest,
+	RateLimitRequestHandler,
+	LegacyStore,
+	Store,
+	IncrementResponse,
+} from './types.js'
+import MemoryStore from './memory-store.js'
+
+/**
+ * Type guard to check if a store is callbacky store.
+ *
+ * @param store {LegacyStore | Store} - The store to check
+ *
+ * @return {boolean} - Whether the store is a callbacky store
+ */
+const isLegacyStore = (store: LegacyStore | Store): store is LegacyStore =>
+	typeof (store as LegacyStore).incr === 'function'
+
+/**
+ * Converts a callbacky store to the promisified version.
+ *
+ * @param store {LegacyStore | Store} - The callbacky store or even a modern store
+ *
+ * @returns {Store} - The promisified version of the store
+ */
+const promisifyStore = (passedStore: LegacyStore | Store): Store => {
+	if (!isLegacyStore(passedStore)) {
+		// It's not an old store, return as is
+		return passedStore
+	}
+
+	// Why can't Typescript understand this?
+	const legacyStore = passedStore
+
+	class PromisifiedStore implements Store {
+		async increment(key: string): Promise<IncrementResponse> {
+			return new Promise((resolve, reject) => {
+				legacyStore.incr(
+					key,
+					(
+						error: Error | undefined,
+						totalHits: number,
+						resetTime: Date | undefined,
+					) => {
+						if (error) reject(error)
+						resolve({ totalHits, resetTime })
+					},
+				)
+			})
+		}
+
+		async decrement(key: string): Promise<void> {
+			return Promise.resolve(legacyStore.decrement(key))
+		}
+
+		async resetKey(key: string): Promise<void> {
+			return Promise.resolve(legacyStore.resetKey(key))
+		}
+
+		async resetAll(): Promise<void> {
+			if (typeof legacyStore.resetAll === 'function')
+				return Promise.resolve(legacyStore.resetAll())
+		}
+	}
+
+	return new PromisifiedStore()
+}
+
+/**
+ * Adds the defaults for options the user has not specified.
+ *
+ * @param options {Options} - The options the user specifies
+ *
+ * @returns {Options} - A complete configuration object
+ */
+const parseOptions = (passedOptions: Partial<Options>): Options => {
+	// Now add the defaults for the other options
+	const options: Options = {
+		windowMs: 60 * 1000,
+		store: new MemoryStore(passedOptions.windowMs ?? 60 * 1000),
+		max: 5,
+		message: 'Too many requests, please try again later.',
+		statusCode: 429,
+		legacyHeaders: true,
+		standardHeaders: false,
```

Comment:
```suggestion
		// support the v5.x names (headers and draft_polli_ratelimit_headers)
		// if the v6.x name is also set, it will override the v5.x names
		legacyHeaders: passedOptions.headers ?? true,
		standardHeaders: passedOptions.draft_polli_ratelimit_headers ?? false,
```

I'm 80% sure this will do the right thing.

Reply:
Doing this will effectively mean that both the old (`headers`, `draft_polli_ratelimit_headers`) and new (`legacyHeaders`, `standardHeaders`) options are valid and we will have two sets of options that do the same thing.

Since this is anyways a major version bump, I think it should be ok if we mark those as deprecated and throw an error if they are passed instead of still supporting them.

Reply:
yep, we write to README, migration guide like breaking changes

Reply:
> Doing this will effectively mean that both the old (headers, draft_polli_ratelimit_headers) and new (legacyHeaders, standardHeaders) options are valid and we will have two sets of options that do the same thing.

Yea, exactly so. And, while that's not great, I do think it's preferable to making a breaking change that doesn't need to be a breaking change. It's about empathy for the end users: we're the ones who decided to rename these things - everyone else just installed the library and followed the directions correctly. (Or, in many cases, someone who doesn't even work there any more installed and configured it, so the new person would be starting from scratch.)

Also, while I like clean code, it's literally a 2-line change to continue supporting the old names in addition to the new ones. This thing gets downloaded 300k times per week - think about how much time it takes for us to not break this, vs the time it would take for everyone else to change their code just to work with our new release.

We'll call it out in the changelog as a non-breaking change, and update the readme to document the new names and mark the old ones as deprecated.

Reply:
> Think about how much time it takes for us to not break this, vs the time it would take for everyone else to change their code just to work with our new release

I agree 👍🏾 

I'll mark the old ones as deprecated, but still support them as suggested.